### PR TITLE
CalculationsShouldNotOverflow.SyntaxKindWalker reduce allocations and evaluations in the hot path

### LINKED ISSUE #8103: CalculationsShouldNotOverflow.SyntaxKindWalker reduce allocations and evaluations in the hot path
There are two things wrong with the walker:

https://github.com/SonarSource/sonar-dotnet/blob/ba330205b4d9cc68f34a4dc594fff8267494c157/analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/CalculationsShouldNotOverflow.cs#L50-L65

### Flawed logic

IsUnchecked is reset too early here/at the wrong position or the IsUnchecked field is not needed at all. It should be reset after the unchecked block more like so:

```cs
        public override void Visit(SyntaxNode node)
        {
            if (!IsUnchecked && !HasOverflow)
            {
                HasOverflow = node.IsAnyKind(
                    SyntaxKind.AddExpression,
                    SyntaxKind.AddAssignmentExpression,
                    SyntaxKind.MultiplyExpression,
                    SyntaxKind.MultiplyAssignmentExpression,
                    SyntaxKind.SubtractExpression,
                    SyntaxKind.SubtractAssignmentExpression,
                    SyntaxKind.PostDecrementExpression,
                    SyntaxKind.PostIncrementExpression,
                    SyntaxKind.PreDecrementExpression,
                    SyntaxKind.PreIncrementExpression);
                base.Visit(node);
            }
        }

        public override void VisitCheckedExpression(CheckedExpressionSyntax node) // same for VisitCheckedStatement
        {
            IsUnchecked = node.IsKind(SyntaxKind.UncheckedExpression); // or dont' call base.VisitCheckedExpression
            base.VisitCheckedExpression(node);
            IsUnchecked = false; // this should actually be [CompilationOptions.CheckOverflow](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilationoptions.checkoverflow?view=roslyn-dotnet-4.6.0)
        }
```

Nested checked/unchecked blocks are potentially wrong (for this to work, IsUnchecked must be a counter).

### Array allocation in the hot path

Visit is performance-sensitive and allocates a params array on each invocation. The IsAnyKind should be replaced by a `node.Kind() is SyntaxKind.AddExpression or SyntaxKind.AddAssignmentExpression or ...` expression:

image

Also affected:

* SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks.CSharp.ConditionEvaluatesToConstant+SyntaxKindWalker.Visit
*  SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks.CSharp.NullPointerDereference+SyntaxKindWalker.Visit

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #8103

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/CalculationsShouldNotOverflow.cs [hunk 1]
```diff
@@ -40,29 +40,64 @@ public override bool ShouldExecute()
...
             }
+            base.Visit(node);
+        }
+
+        public override void VisitBinaryExpression(BinaryExpressionSyntax node)
+        {
+            HasOverflow |= node.Kind() is
+                SyntaxKind.AddExpression or
+                SyntaxKind.MultiplyExpression or
+                SyntaxKind.SubtractExpression;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/CalculationsShouldNotOverflow.cs:62]: Can you put these on a single line? Othwerwise, the `or` would need to be on the beginning of next line, similar to `||`

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    [DataRow("""
+        i += 1;
+        i /= 1;
+        """)]
+    public void HasOverflowExpressions_True(string statement)
+    {
+        var (tree, _) = TestHelper.CompileCS(WrapInMethod($$"""
+            var i = 0;
+            {{statement}}
+            """));
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs:61]: Same below

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        _ = i++;
+        _ = i!;
+        """)]
+    [DataRow("""
+        i += 1;
+        i /= 1;
+        """)]
+    public void HasOverflowExpressions_True(string statement)
+    {
+        var (tree, _) = TestHelper.CompileCS(WrapInMethod($$"""
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs:58]: Should be enough here
```suggestion
        var (tree, _) = TestHelper.CompileCS(WrapInMethod($"""
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Minor polishing

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        sut.SafeVisit(tree.GetRoot());
+        sut.Should().BeEquivalentTo(new
+        {
+            HasOverflow = expectedHasOverflow,
+            IsUnchecked = expectedIsUnchecked,
+        });
+    }
+
+    [DataTestMethod]
+    [DataRow("_ = 1 + 1;", true)]            // Fix: should be false as it is checked at compile time
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs:215]: These are all cases with a binary operation but never an overflow. This is the case for constants (checked by the compiler) and string concatenation. The idea was to document the cases here and add an issue to fix later. But I wanted to know if there was enough value in adding it.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            _ = i + i;
+        }
+        """)]
+    [DataRow("_ = unchecked(i + i);")]
+    public void HasOverflowExpressions_IsUnchecked_True(string statement)
+    {
+        var (tree, _) = TestHelper.CompileCS(WrapInMethod($$"""
+            var i = 0;
+            {{statement}}
+            """));
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs]: I put all local declarations on the same line but kept the statement separately.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs:202]: Unless we have FPs reported, I would not add issue for that.
Let's remove the `Fix:` prefix, the `Should ....` is enough to explain the problem

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    [DataRow("i %= 1;")]
+    public void HasOverflowExpressions_False(string statement)
+    {
+        var tree = TestHelper.CompileCS(WrapInMethod($"""
+            var i = 0; var b = true;
+            {statement}
+            """)).Tree;
+        var sut = new CalculationsShouldNotOverflow.SyntaxKindWalker();
+        sut.SafeVisit(tree.GetRoot());
+        sut.HasOverflow.Should().Be(false);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs:98]: There're dedicated methods for true and false. Same below.
```suggestion
        sut.HasOverflow.Should().BeFalse();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    [DataRow("_ = i - 0;", true)]            // Fix: should be false as one operand is the identity element of subtraction
+    [DataRow("_ = i * 1;", true)]            // Fix: should be false as one operand is the identity element of multiplication
+    [DataRow("_ = i * 0;", true)]            // Fix: should be false as the result is always 0
+    [DataRow("""_ = "" + "";""", true)]      // Fix: should be false as both sides are string literals
+    [DataRow("""_ = s + "";""", true)]       // Fix: should be false as right side is a string literal
+    [DataRow("""_ = "" + s;""", true)]       // Fix: should be false as left side is a string literal
+    [DataRow("""_ = "" + s + s;""", true)]   // Fix: should be false as one of the operands is a string literal
+    [DataRow("""_ = s + "" + s;""", true)]   // Fix: should be false as one of the operands is a string literal
+    [DataRow("""_ = s + s + "";""", true)]   // Fix: should be false as one of the operands is a string literal
+    public void HasOverflowExpressions_Literals(string statement, bool expected)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs:217]: I would remove the `bool expected`, as it's always true. If `False` is needed, second method can be added, or parameter can be re-added later.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs (hunk 7 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/CalculationsShouldNotOverflowSyntaxKindWalkerTest.cs:217]: This misses a bit the purpose of the tests: Enable an incremental optimization by turning these into `false`. I kept it this way.

**Pull request merged**